core/src/main/scala/kafka/coordinator/group/GroupMetadata.scala

```diff
@@ -434,18 +441,54 @@ private[group] class GroupMetadata(val groupId: String, initialState: GroupState
     }.toMap
   }
 
-  def removeExpiredOffsets(startMs: Long) : Map[TopicPartition, OffsetAndMetadata] = {
-    val expiredOffsets = offsets
-      .filter {
-        case (topicPartition, commitRecordMetadataAndOffset) =>
-          commitRecordMetadataAndOffset.offsetAndMetadata.expireTimestamp < startMs && !pendingOffsetCommits.contains(topicPartition)
-      }
-      .map {
-        case (topicPartition, commitRecordOffsetAndMetadata) =>
-          (topicPartition, commitRecordOffsetAndMetadata.offsetAndMetadata)
-      }
+  def removeExpiredOffsets(currentTimestamp: Long, offsetRetentionMs: Long) : Map[TopicPartition, OffsetAndMetadata] = {
+    val expiredOffsets: Map[TopicPartition, CommitRecordMetadataAndOffset] = protocolType match {
+      case Some(protocol) if is(Empty) =>
+        // no consumer exists in the group =>
+        // if retention period has passed since group became Empty, expire all offsets with no pending offset commit
+        offsets.filter {
+          case (topicPartition, commitRecordMetadataAndOffset) =>
+            !pendingOffsetCommits.contains(topicPartition) && {
+              commitRecordMetadataAndOffset.offsetAndMetadata.expireTimestamp match {
+                case None =>
+                  // current version with no per partition retention
+                  currentTimestamp - currentStateTimestamp >= offsetRetentionMs
+                case Some(expireTimestamp) =>
+                  // older versions with explicit expire_timestamp field => old expiration semantics is used
+                  currentTimestamp >= expireTimestamp
+              }
+            }
+        }.toMap
+
+      case Some(protocol) =>
+        //TODO (KIP-211): expire offsets of partitions the group is no longer subscribed to, assuming retention period
```

Comment:
We can post a JIRA to keep track of this rather than doing it in the code. For now, maybe we can just remove this case?

Reply:
Sure, it looks like a `case _ => Map()` is required at the end to cover the whole spectrum.
I created [KAFKA-6951](https://issues.apache.org/jira/browse/KAFKA-6951) to keep track of this portion.